src/Async/Async.spec.js

```diff
@@ -1253,3 +1253,53 @@ test('Async chain properties (Monad)', t => {
 
   t.end()
 })
+
+test('Async applyTo', t => {
+  const f = sinon.spy()
+
+  Async.of(3).applyTo(Async.of())
+
+  t.ok(f.calledWith(0), 'wraps the value passed into an Async')
+
+  t.end()
+})
+
+// todo: What can we generalize for all Applicative law tests??
```

Comment:
Was thinking about setting something up for all law specs.
So it would be something like as a support file in the specs:
```js
exports.setoid = {
  reflexivity: a => a.equals(a),
  symmetry: (a, b) => a.equals(b) === b.equals(a),
  transitivity: (a, b, c) => !(a.equals(b) && b.equals(c)) || a.equals(c),
}
 exports.semigroup = {
  associativity: (a, b, c) =>
    equals(a.concat(b).concat(c), a.concat(b.concat(c))),
}
 exports.monoid = {
  rightIdentity: (type, m) => equals(m.concat(type.empty()), m),
  leftIdentity: (type, m) => equals(type.empty().concat(m), m),
}
```
Then have another support file that checks for existing methods (like you need `map` for `ap`, etc).
If you would like to add the start  (just for `Apply` and `Applicative`) that would be cool, if not just mimic how `Apply` and `Applicative` laws are tested for non-FL for now, then we can do a sweep and clean up all the other specs in a separate PR.

Reply:
So I am thinking the laws file would look something like this:
```js
const fl = require('../core/flNames')

const compose =
  f => g => x => f(g(x))

const identity =
  x => x

const applyTo =
  x => f => f(x)

module.exports = {
  apply: {
    composition: (equals, f, g, m) => equals(
      f.map(compose).ap(g).ap(m),
      f.ap(g.ap(m))
    )
  },

  'fl/apply': {
    composition: (equals, f, g, m) => equals(
      m[fl.ap](g[fl.ap](f[fl.map](compose))),
      m[fl.ap](g)[fl.ap](f)
    )
  },

  applicative: type => ({
    identity: (equals, x) => equals(
      type.of(identity).ap(x),
      x
    ),

    homomorphism: (equals, f, x) => equals(
      type.of(f).ap(type.of(x)),
      type.of(f(x))
    ),

    interchange: (equals, mFn, x) => equals(
      mFn.ap(type.of(x)),
      type.of(applyTo(x)).ap(mFn)
    )
  })
}
```
providing the `equals` function on each test for those cases that we do not have a `Setoid` (`Async` is not a `Setoid` in `crocks`. That way we can set up a `t.plan(n)` and use a function like:
```js

  const spec = name => (a, b) => {
    const aRes = sinon.spy()
    const bRes = sinon.spy()

    a.fork(unit, aRes)
    b.fork(unit, bRes)

    t.same(aRes.args[0], bRes.args[0], name)
  }
```

for testing equality on all laws in the `Async` file. So the `Apply` laws (for non-fantasyland) would look like:
```js
test('Async ap properties (Apply)', t => {
  t.plan(3)

  t.ok(isFunction(Async(unit).map), 'implements the Functor spec')
  t.ok(isFunction(Async(unit).ap), 'provides an ap function')

  const times10 = x => x * 10
  const plus10 = x => x + 10

  laws.apply.composition(
    spec('composition'),
    Async.of(times10),
    Async.of(plus10),
    Async.of(0)
  )
})
```

But for cases like `Maybe` that are indeed `Setoid` and not lazy, etc, etc then it just looks like:
```js

test('Maybe ap properties (Apply)', t => {
  const j = Maybe.Just(3)
  const n = Maybe.Nothing()

  t.ok(isFunction(j.ap), 'Just provides an ap function')
  t.ok(isFunction(j.map), 'Just implements the Functor spec')

  t.ok(isFunction(n.ap), 'Nothing provides an ap function')
  t.ok(isFunction(n.map), 'Nothing implements the Functor spec')

  const f = x => x * 10
  const g = x => x + 10

  t.ok(laws.apply.composition(equals, Maybe.of(f), Maybe.of(g), j), 'composition Just')
  t.ok(laws.apply.composition(equals, Maybe.of(f), Maybe.of(g), n), 'composition Nothing')

  t.end()
})
```

But I am still not 💯 on that API tho.

Reply:
That makes sense, I like the `laws.{type}.{law_name}` scheme 👍 

I'll add the Apply/Applicative laws like you recommended in this PR, then we can add others in a follow

Reply:
Awesome! Thanks for doing this, Super exciting!